compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -0,0 +1,886 @@
+    }
+  }
+
+  /** Resolve C.this that appear in `klass` */
+  def resolveThis(target: ClassSymbol, thisV: Value, klass: ClassSymbol, source: Tree): Contextual[Value] = log("resolving " + target.show + ", this = " + thisV.show + " in " + klass.show, printer, res => res.asInstanceOf[Value].show) {
+    if target == klass then thisV
+    else if target.is(Flags.Package) || target.isStaticOwner && target != heap.klass then Hot
+    else
+      thisV match
+        case Hot | _: ThisRef => Hot
+        case warm: Warm =>
+          // use existing type information as a shortcut
+          val tref = typeRefOf(warm.klass.typeRef.baseType(klass))
+          if tref.prefix == NoPrefix then
+            // Current class is local, in the enclosing scope of `warm.klass`
+            val outerCls = warm.klass.owner.enclosingClass.asClass
+            resolveThis(target, warm.outer, outerCls, source)
+          else
+            val outerCls = klass.owner.enclosingClass.asClass
+            val warmOuterCls = warm.klass.owner.enclosingClass.asClass
+            val res = cases(tref.prefix, warm.outer, warmOuterCls, source)
+            assert(res.errors.isEmpty, "unexpected error " + res)
+            resolveThis(target, res.value, outerCls, source)
+        case _ =>
+          // report.error("unexpected thisV = " + thisV + ", target = " + target.show + ", klass = " + klass.show, source.srcPos)
+          Cold
+  }
+
+  /** Compute the outer value that correspond to `tref.prefix` */
+  def outerValue(tref: TypeRef, thisV: Value, klass: ClassSymbol, source: Tree): Contextual[Result] =
+    val cls = tref.classSymbol.asClass
+    if tref.prefix == NoPrefix then
+      val enclosing = cls.owner.lexicallyEnclosingClass.asClass
+      val outerV = resolveThis(enclosing, thisV, klass, source)
+      Result(outerV, noErrors)
+    else
+      cases(tref.prefix, thisV, klass, source)
+
+  /** Initialize part of an abstract object in `klass` of the inheritance chain */
+  def init(tpl: Template, thisV: ThisRef | Warm, klass: ClassSymbol): Contextual[Result] = log("init " + klass.show, printer, res => res.asInstanceOf[Result].show) {
+    val errorBuffer = new mutable.ArrayBuffer[Error]
+
+    // init param fields
+    klass.paramAccessors.foreach { acc =>
+      if (!acc.is(Flags.Method)) {
+        printer.println(acc.show + " initialized")
+        thisV.updateField(acc, Hot)
+      }
+    }
+
+    def superCall(tref: TypeRef, ctor: Symbol, source: Tree): Unit =
+      val cls = tref.classSymbol.asClass
+      // update outer for super class
+      // ignored as they are all hot
+
+      // follow constructor
+      if !cls.defTree.isEmpty then
+        val res2 = thisV.call(ctor, superType = NoType, source)(using heap, ctx, trace.add(source))
+        errorBuffer ++= res2.errors
+
+    // parents
+    def initParent(parent: Tree) = parent match {
+      case tree @ Block(stats, NewExpr(tref, New(tpt), ctor, argss)) =>
```

Comment:
I don't fully understand these cases of parent. Perhaps comment like // class A extends B(arg)